bottlerocket/agents/src/sonobuoy.rs

```diff
@@ -81,19 +92,116 @@ pub async fn rerun_failed_sonobuoy(
         .args(kubeconfig_arg.to_owned())
         .arg("run")
         .args(e2e_repo_arg)
-        .arg("--wait")
         .arg("--rerun-failed")
         .arg(results_filepath.as_os_str())
         .status()
         .context(error::SonobuoyProcessSnafu)?;
-    info!("Sonobuoy testing has completed, checking results");
 
     // TODO - log something or check what happened?
     ensure!(status.success(), error::SonobuoyRunSnafu);
 
+    info!("Sonobuoy testing has started, waiting for status to be available");
+    tokio::time::timeout(
+        Duration::from_secs(300),
+        wait_for_sonobuoy_status(kubeconfig_path, None),
+    )
+    .await
+    .context(error::SonobuoyTimeoutSnafu)??;
+    info!("Sonobuoy status is available, waiting for test to complete");
+    wait_for_sonobuoy_results(kubeconfig_path, None).await?;
+    info!("Sonobuoy testing has completed, checking results");
+
     results_sonobuoy(kubeconfig_path, results_dir)
 }
 
+pub async fn wait_for_sonobuoy_status(
+    kubeconfig_path: &str,
+    namespace: Option<&str>,
+) -> Result<(), error::Error> {
+    loop {
+        let kubeconfig_arg = vec!["--kubeconfig", kubeconfig_path];
+        let namespace_arg = namespace
+            .map(|namespace| vec!["--namespace", namespace])
+            .unwrap_or_default();
+
+        trace!("Checking test status");
+        let run_result = Command::new("/usr/bin/sonobuoy")
+            .args(kubeconfig_arg)
+            .args(namespace_arg)
+            .arg("status")
+            .arg("--json")
+            .output()
+            .context(error::SonobuoyProcessSnafu)?;
+
+        let stdout = String::from_utf8_lossy(&run_result.stdout);
+        info!("Parsing the following sonobuoy results output:\n{}", stdout);
```

Comment:
You should check the exit code before parsing stdout.

And when you do, is there any chance of a transient error leading to a non-zero exit?

As I'm looking at all the `?` operators, I'm wondering how likely it is that we can error out leaving the cluster in a bad state because it has sonobuoy still running. Can we have some best effort teardown that occurs before the agent is about exit non-zero or something?

Reply:
If the agent errors then `terminate_best_effort` should be called which will cal `sonobuoy delete`.

Reply:
> If the agent errors then terminate_best_effort should be called which will cal sonobuoy delete.

Oh good, I forgot how it works. The main thing is we want to make sure we don't make it less reliable by having more calls that could (transiently?) exit non-zero or error at `?`. I guess we can revisit if we encounter issues.